### PR TITLE
Refactor busybox downloading as generic "frozen-images"

### PR DESCRIPTION
**Pull request by tianon**
This makes it much simpler to add new "frozen" images -- simply add them to the `Dockerfile` and in `hack/make/.ensure-frozen-images` and you're off to the races.

### CODE REVIEW

**Comment by tianon** [comment]
As a concrete example, #10893 could be implemented with this as:
```patch
diff --git a/Dockerfile b/Dockerfile
index 412fff4..bbbea17 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -145,7 +145,8 @@ RUN ln -sfv $PWD/.bashrc ~/.bashrc
 # Get useful and necessary Hub images so we can "docker load" locally instead of pulling
 COPY contrib/download-frozen-image.sh /go/src/github.com/docker/docker/contrib/
 RUN ./contrib/download-frozen-image.sh /docker-frozen-images \
-       busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125
+       busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 \
+       nginx:latest@2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d
 # see also "hack/make/.ensure-frozen-images" (which needs to be updated any time this list is)
 
 # Install man page generator
diff --git a/project/make/.ensure-frozen-images b/project/make/.ensure-frozen-images
index ee3e92b..d9ffc41 100644
--- a/project/make/.ensure-frozen-images
+++ b/project/make/.ensure-frozen-images
@@ -4,6 +4,7 @@ set -e
 # this list should match roughly what's in the Dockerfile (minus the explicit image IDs, of course)
 images=(
        busybox:latest
+       nginx:latest
 )
 
 if ! docker inspect "${images[@]}" &> /dev/null; then
```

**Comment by tianon** [comment]
Example output for the patch in my previous comment:
```
Step 33 : RUN ./contrib/download-frozen-image.sh /docker-frozen-images 	busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 	nginx:latest@2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d
 ---> Running in 100c0a859a31
Downloading 'busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125' (4 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%

Downloading 'nginx:latest@2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d' (13 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Download of images into '/docker-frozen-images' complete.
Use something like the following to load the result into a Docker daemon:
  tar -cC '/docker-frozen-images' . | docker load
 ---> c69759b6f7b4
Removing intermediate container 100c0a859a31
```

**Comment by tianon** [comment]
Want to add the v1 registry image too, to cover #10967?  No problem:
```patch
diff --git a/Dockerfile b/Dockerfile
index 412fff4..5001eaf 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -145,7 +145,9 @@ RUN ln -sfv $PWD/.bashrc ~/.bashrc
 # Get useful and necessary Hub images so we can "docker load" locally instead of pulling
 COPY contrib/download-frozen-image.sh /go/src/github.com/docker/docker/contrib/
 RUN ./contrib/download-frozen-image.sh /docker-frozen-images \
-       busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125
+       busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 \
+       nginx:latest@2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d \
+       registry:latest@c55308716b3671d8a238a6c1245a60f66d76a3e1404b7b8b6e5fe0e65bae4943
 # see also "hack/make/.ensure-frozen-images" (which needs to be updated any time this list is)
 
 # Install man page generator
diff --git a/project/make/.ensure-frozen-images b/project/make/.ensure-frozen-images
index ee3e92b..fb5b635 100644
--- a/project/make/.ensure-frozen-images
+++ b/project/make/.ensure-frozen-images
@@ -4,6 +4,8 @@ set -e
 # this list should match roughly what's in the Dockerfile (minus the explicit image IDs, of course)
 images=(
        busybox:latest
+       nginx:latest
+       registry:latest
 )
 
 if ! docker inspect "${images[@]}" &> /dev/null; then
```

And output:
```
Step 32 : COPY contrib/download-frozen-image.sh /go/src/github.com/docker/docker/contrib/
 ---> Using cache
 ---> 2caa666227e9
Step 33 : RUN ./contrib/download-frozen-image.sh /docker-frozen-images 	busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 	nginx:latest@2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d 	registry:latest@c55308716b3671d8a238a6c1245a60f66d76a3e1404b7b8b6e5fe0e65bae4943
 ---> Running in 256ed5735e11
Downloading 'busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125' (4 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%

Downloading 'nginx:latest@2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d' (13 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'registry:latest@c55308716b3671d8a238a6c1245a60f66d76a3e1404b7b8b6e5fe0e65bae4943' (15 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Download of images into '/docker-frozen-images' complete.
Use something like the following to load the result into a Docker daemon:
  tar -cC '/docker-frozen-images' . | docker load
 ---> 038ec37e7563
Removing intermediate container 256ed5735e11
```

File: project/make/test-integration-cli [hunk 1]
```diff
@@ -16,7 +16,7 @@ bundle_test_integration_cli() {
 	# even and especially on test failures
 	didFail=
 	if ! {
-		source "$(dirname "$BASH_SOURCE")/.ensure-busybox"
+		source "$(dirname "$BASH_SOURCE")/.ensure-frozen-images"
 		source "$(dirname "$BASH_SOURCE")/.ensure-emptyfs"
```
Comment (tiborvass) [inline project/make/test-integration-cli]: @tianon possibly stupid question, is it deliberate that we put emptyfs after our images?

Comment (tianon) [inline project/make/test-integration-cli]: Definitely not a stupid question -- the answer is that it's pretty arbitrary.  Back when it was `scratch` and we were building `busybox` from source, it actually mattered, but now it doesn't really matter much at all.

**Comment by tianon** [comment]
Want every single current tag of `debian` all at once, like a true and proper masochist?  Easy:
```patch
diff --git a/Dockerfile b/Dockerfile
index 412fff4..fe339eb 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -145,7 +145,25 @@ RUN ln -sfv $PWD/.bashrc ~/.bashrc
 # Get useful and necessary Hub images so we can "docker load" locally instead of pulling
 COPY contrib/download-frozen-image.sh /go/src/github.com/docker/docker/contrib/
 RUN ./contrib/download-frozen-image.sh /docker-frozen-images \
-	busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125
+	busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 \
+	debian:8.0@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce \
+	debian:8@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce \
+	debian:jessie@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce \
+	debian:oldstable@5971404fdb2ed835e86619996e02719270e9a10bd1232ce40daeb31e5b5a0cfd \
+	debian:sid@c19cc6c3dedbc52971ef56e0612d57ba4d689c8cf65de31600374ba5806ae6ed \
+	debian:6.0.10@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 \
+	debian:6.0@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 \
+	debian:6@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 \
+	debian:squeeze@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 \
+	debian:stable@928892d9c89c3cf9737a90ae405c1bb620070c78bdef9775b6c833efd015e642 \
+	debian:testing@3c3575a11f159b5fa77d58ea52146d824b3208589fc3020710f5087bfa6c1940 \
+	debian:unstable@ca405a8eedaa641da5cb41d3dc65ee82b1b6b32a6d3e2ffc4120c5ed328d2315 \
+	debian:7.8@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 \
+	debian:7@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 \
+	debian:wheezy@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 \
+	debian:latest@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 \
+	debian:rc-buggy@52835f909e3daf22b061df15e5a437d67151de148464368aaad730d21632f660 \
+	debian:experimental@5b49567f7cadb6405bc3acfcf81dad31df186aa2b4f486d7ab644757c93eeb47
 # see also "hack/make/.ensure-frozen-images" (which needs to be updated any time this list is)
 
 # Install man page generator
diff --git a/project/make/.ensure-frozen-images b/project/make/.ensure-frozen-images
index ee3e92b..6cd4316 100644
--- a/project/make/.ensure-frozen-images
+++ b/project/make/.ensure-frozen-images
@@ -4,6 +4,24 @@ set -e
 # this list should match roughly what's in the Dockerfile (minus the explicit image IDs, of course)
 images=(
 	busybox:latest
+	debian:8.0
+	debian:8
+	debian:jessie
+	debian:oldstable
+	debian:sid
+	debian:6.0.10
+	debian:6.0
+	debian:6
+	debian:squeeze
+	debian:stable
+	debian:testing
+	debian:unstable
+	debian:7.8
+	debian:7
+	debian:wheezy
+	debian:latest
+	debian:rc-buggy
+	debian:experimental
 )
 
 if ! docker inspect "${images[@]}" &> /dev/null; then
```

And output:
```
Step 33 : RUN ./contrib/download-frozen-image.sh /docker-frozen-images 	busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125 	debian:8.0@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce 	debian:8@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce 	debian:jessie@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce 	debian:oldstable@5971404fdb2ed835e86619996e02719270e9a10bd1232ce40daeb31e5b5a0cfd 	debian:sid@c19cc6c3dedbc52971ef56e0612d57ba4d689c8cf65de31600374ba5806ae6ed 	debian:6.0.10@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 	debian:6.0@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 	debian:6@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 	debian:squeeze@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873 	debian:stable@928892d9c89c3cf9737a90ae405c1bb620070c78bdef9775b6c833efd015e642 	debian:testing@3c3575a11f159b5fa77d58ea52146d824b3208589fc3020710f5087bfa6c1940 	debian:unstable@ca405a8eedaa641da5cb41d3dc65ee82b1b6b32a6d3e2ffc4120c5ed328d2315 	debian:7.8@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 	debian:7@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 	debian:wheezy@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 	debian:latest@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444 	debian:rc-buggy@52835f909e3daf22b061df15e5a437d67151de148464368aaad730d21632f660 	debian:experimental@5b49567f7cadb6405bc3acfcf81dad31df186aa2b4f486d7ab644757c93eeb47
 ---> Running in 73a0d9084867
Downloading 'busybox:latest@4986bf8c15363d1c5d15512d5266f8777bfba4974ac56e3270e7760f6f0a8125' (4 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%
######################################################################## 100.0%

Downloading 'debian:8.0@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:8@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce' (3 layers)...
skipping existing 0e30e84e9513
skipping existing 8771fbfe935c
skipping existing 511136ea3c5a

Downloading 'debian:jessie@0e30e84e9513cb3d9d991389adfb77ac6afc9d75a5234e3d61af2f080c75d0ce' (3 layers)...
skipping existing 0e30e84e9513
skipping existing 8771fbfe935c
skipping existing 511136ea3c5a

Downloading 'debian:oldstable@5971404fdb2ed835e86619996e02719270e9a10bd1232ce40daeb31e5b5a0cfd' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:sid@c19cc6c3dedbc52971ef56e0612d57ba4d689c8cf65de31600374ba5806ae6ed' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:6.0.10@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:6.0@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873' (3 layers)...
skipping existing 1df4e3a3dcf7
skipping existing 785f3f1395b4
skipping existing 511136ea3c5a

Downloading 'debian:6@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873' (3 layers)...
skipping existing 1df4e3a3dcf7
skipping existing 785f3f1395b4
skipping existing 511136ea3c5a

Downloading 'debian:squeeze@1df4e3a3dcf791d41af93d5c7763fdc8bc60ef2065ddf075285fec9586589873' (3 layers)...
skipping existing 1df4e3a3dcf7
skipping existing 785f3f1395b4
skipping existing 511136ea3c5a

Downloading 'debian:stable@928892d9c89c3cf9737a90ae405c1bb620070c78bdef9775b6c833efd015e642' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:testing@3c3575a11f159b5fa77d58ea52146d824b3208589fc3020710f5087bfa6c1940' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:unstable@ca405a8eedaa641da5cb41d3dc65ee82b1b6b32a6d3e2ffc4120c5ed328d2315' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:7.8@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444' (3 layers)...
######################################################################## 100.0%
######################################################################## 100.0%
skipping existing 511136ea3c5a

Downloading 'debian:7@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444' (3 layers)...
skipping existing c90d655b99b2
skipping existing 30d39e59ffe2
skipping existing 511136ea3c5a

Downloading 'debian:wheezy@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444' (3 layers)...
skipping existing c90d655b99b2
skipping existing 30d39e59ffe2
skipping existing 511136ea3c5a

Downloading 'debian:latest@c90d655b99b2ec5b7e94d38c87f92dce015c17a313caeaae0e980d9b9bed8444' (3 layers)...
skipping existing c90d655b99b2
skipping existing 30d39e59ffe2
skipping existing 511136ea3c5a

Downloading 'debian:rc-buggy@52835f909e3daf22b061df15e5a437d67151de148464368aaad730d21632f660' (4 layers)...
######################################################################## 100.0%
skipping existing c19cc6c3dedb
skipping existing 981ccdc96d17
skipping existing 511136ea3c5a

Downloading 'debian:experimental@5b49567f7cadb6405bc3acfcf81dad31df186aa2b4f486d7ab644757c93eeb47' (4 layers)...
######################################################################## 100.0%
skipping existing ca405a8eedaa
skipping existing 3e972c473e06
skipping existing 511136ea3c5a

Download of images into '/docker-frozen-images' complete.
Use something like the following to load the result into a Docker daemon:
  tar -cC '/docker-frozen-images' . | docker load
 ---> 5a28887d6130
Removing intermediate container 73a0d9084867
```

**Comment by tianon** [comment]
Note all the "skipping existing" bits there in that `debian` example (that we'd never actually merge, it's just a great illustration) -- any layers we've got from other images are only downloaded the first time. :+1:

**Comment by tiborvass** [comment]
In your debian example, it'd be great to have one source of truth e.g. in a file that both the RUN command and the .ensure script uses. But since we only need busybox for not sure it matters :)

**Comment by tianon** [comment]
Yeah, I thought about scraping the `Dockerfile` itself for the list.  I can adapt to do that, but it's going to be pretty hacky.

**Comment by tianon** [comment]
I also want to avoid too many `COPY` instructions before `RUN` instructions since something as small as an mtime change will bust the cache there.

**Comment by tiborvass** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Comment by unclejack** [comment]
LGTM

**Pull request merged**